File: qt-easy-tier-master/SRC/backend/OneClickConnectionCode.cpp
//
// Created by Copilot on 2026/5/13.
//

#include "backend/OneClickConnectionCode.h"

#include <QRegularExpression>

#include <random>

namespace {

static const QString PRIMARY_BASE32_CHARS = QStringLiteral("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567");
static const QString LEGACY_BASE32_CHARS = QStringLiteral("ABCDEFGHJKLMNPQRSTUVWXYZ23456789");

} // namespace

namespace EasyTierBackend {

QString base32Encode(const QByteArray &data, const QString &alphabet = PRIMARY_BASE32_CHARS)
{
    QString result;
    int buffer = 0;
    int bitsLeft = 0;

    for (int i = 0; i < static_cast<int>(data.size()); ++i) {
        buffer = (buffer << 8) | static_cast<unsigned char>(data[i]);
        bitsLeft += 8;

        while (bitsLeft >= 5) {
            result += alphabet[(buffer >> (bitsLeft - 5)) & 0x1F];
            bitsLeft -= 5;
        }
    }

    if (bitsLeft > 0) {
        result += alphabet[(buffer << (5 - bitsLeft)) & 0x1F];
    }

    return result;
}

QByteArray base32Decode(const QString &encoded, const QString &alphabet)
{
    QByteArray result;
    int buffer = 0;
    int bitsLeft = 0;

    for (int i = 0; i < static_cast<int>(encoded.length()); ++i) {
        const QChar ch = encoded[i];
        if (ch == QLatin1Char('=')) {
            continue;
        }

        const int val = static_cast<int>(alphabet.indexOf(ch.toUpper()));
        if (val == -1) {
            continue;
        }

        buffer = (buffer << 5) | val;
        bitsLeft += 5;

        if (bitsLeft >= 8) {
            result.append(static_cast<char>((buffer >> (bitsLeft - 8)) & 0xFF));
            bitsLeft -= 8;
        }
    }

    return result;
}

QPair<QString, QString> generateRoomCredentials()
{
    static const std::string charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789*&#$!";
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, static_cast<int>(charset.size() - 1));

    QByteArray networkIdRaw(7, 0);
    networkIdRaw[0] = 'Q';
    networkIdRaw[1] = 'E';
    for (int i = 2; i < 7; ++i) {
        networkIdRaw[i] = charset[dis(gen)];
    }

    QByteArray passwordRaw(8, 0);
    for (int i = 0; i < 8; ++i) {
        passwordRaw[i] = charset[dis(gen)];
    }

    const QString networkId = QStringLiteral("QtET-OneClick-") + QString::fromLatin1(networkIdRaw);
    const QString password = QString::fromLatin1(passwordRaw);

    return qMakePair(networkId, password);
}

QString encodeConnectionCode(const QString &networkId, const QString &password)
{
    QString networkPureId = networkId;
    networkPureId.remove(QStringLiteral("QtET-OneClick-"));

    const QByteArray networkIdData = networkPureId.toLatin1();
    const QByteArray passwordData = password.toLatin1();

    if (networkIdData.size() < 2 || networkIdData[0] != 'Q' || networkIdData[1] != 'E') {
        return {};
    }

    const QString encodedNetworkId = base32Encode(networkIdData);
    const QString encodedPassword = base32Encode(passwordData);

    if (encodedNetworkId.length() != 12 || encodedPassword.length() != 13) {
        return {};
    }

    const QString code = encodedNetworkId.left(4) + "-"
        + encodedNetworkId.mid(4, 4) + "-"
        + encodedNetworkId.mid(8, 4) + "-"
        + encodedPassword.left(5) + "-"
        + encodedPassword.mid(5, 5) + "-"
        + encodedPassword.mid(10, 3);

    return code.toUpper();
}

static QPair<QString, QString> decodeConnectionCodeWithAlphabet(const QString &cleanCode, const QString &alphabet)
{
    const QString encodedNetworkId = cleanCode.left(12);
    const QString encodedPassword = cleanCode.mid(12, 13);

    const QByteArray networkIdData = base32Decode(encodedNetworkId, alphabet);
    const QByteArray passwordData = base32Decode(encodedPassword, alphabet);

    if (networkIdData.size() != 7 || passwordData.size() != 8) {
        return qMakePair(QString(), QString());
    }

    if (networkIdData[0] != 'Q' || networkIdData[1] != 'E') {
        return qMakePair(QString(), QString());
    }

    const QString networkPureId = QString::fromLatin1(networkIdData);
    const QString password = QString::fromLatin1(passwordData);

    return qMakePair(QStringLiteral("QtET-OneClick-") + networkPureId, password);
}

QPair<QString, QString> decodeConnectionCode(const QString &code)
{
    QString cleanCode = code;
    static const QRegularExpression re("[^A-Za-z2-9]");
    cleanCode.remove(re);
    cleanCode = cleanCode.toUpper();

    if (cleanCode.length() != 25) {
        return qMakePair(QString(), QString());
    }

    const auto primaryDecoded = decodeConnectionCodeWithAlphabet(cleanCode, PRIMARY_BASE32_CHARS);
    if (!primaryDecoded.first.isEmpty() && !primaryDecoded.second.isEmpty()) {
        return primaryDecoded;
    }

    return decodeConnectionCodeWithAlphabet(cleanCode, LEGACY_BASE32_CHARS);
}

} // namespace EasyTierBackend

File: backend/src/main/java/com/easytier/backend/OneClickConnectionCode.kt
package com.easytier.backend

import kotlin.random.Random

private const val PRIMARY_BASE32 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
private const val LEGACY_BASE32 = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"

fun base32Encode(data: ByteArray, alphabet: String = PRIMARY_BASE32): String {
    val result = StringBuilder()
    var buffer = 0
    var bitsLeft = 0

    for (byte in data) {
        buffer = (buffer shl 8) or (byte.toInt() and 0xFF)
        bitsLeft += 8
        while (bitsLeft >= 5) {
            result.append(alphabet[(buffer shr (bitsLeft - 5)) and 0x1F])
            bitsLeft -= 5
        }
    }

    if (bitsLeft > 0) {
        result.append(alphabet[(buffer shl (5 - bitsLeft)) and 0x1F])
    }

    return result.toString()
}

private fun base32Decode(encoded: String, alphabet: String): ByteArray {
    val bytes = ArrayList<Byte>()
    var buffer = 0
    var bitsLeft = 0

    for (ch in encoded.uppercase()) {
        if (ch == '=') continue
        val index = alphabet.indexOf(ch)
        if (index < 0) continue

        buffer = (buffer shl 5) or index
        bitsLeft += 5

        if (bitsLeft >= 8) {
            bytes.add(((buffer shr (bitsLeft - 8)) and 0xFF).toByte())
            bitsLeft -= 8
        }
    }

    return bytes.toByteArray()
}

private fun decodeConnectionCodeWithAlphabet(code: String, alphabet: String): Pair<String, String> {
    val encodedNetworkId = code.substring(0, 12)
    val encodedPassword = code.substring(12, 25)
    val networkIdData = base32Decode(encodedNetworkId, alphabet)
    val passwordData = base32Decode(encodedPassword, alphabet)

    if (networkIdData.size != 7 || passwordData.size != 8) return "" to ""
    if (networkIdData[0] != 'Q'.code.toByte() || networkIdData[1] != 'E'.code.toByte()) return "" to ""

    return "QtET-OneClick-" + networkIdData.decodeToString() to passwordData.decodeToString()
}

fun generateRoomCredentials(): Pair<String, String> {
    val charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789*&#$!"
    val networkIdBytes = ByteArray(7).apply {
        this[0] = 'Q'.code.toByte()
        this[1] = 'E'.code.toByte()
        for (index in 2 until 7) {
            this[index] = charset[Random.nextInt(charset.length)].code.toByte()
        }
    }
    val passwordBytes = ByteArray(8) { charset[Random.nextInt(charset.length)].code.toByte() }
    val networkId = "QtET-OneClick-" + networkIdBytes.decodeToString()
    val password = passwordBytes.decodeToString()
    return networkId to password
}

fun encodeConnectionCode(networkId: String, password: String): String {
    val pureId = networkId.removePrefix("QtET-OneClick-")
    val networkIdData = pureId.toByteArray()
    val passwordData = password.toByteArray()

    if (networkIdData.size < 2 || networkIdData[0] != 'Q'.code.toByte() || networkIdData[1] != 'E'.code.toByte()) {
        return ""
    }

    // Qt 端当前仍使用旧字母表编码联机码，Android 房主也按同一套生成，避免 PC 房客侧判“联机码错误”。
    val encodedNetworkId = base32Encode(networkIdData, LEGACY_BASE32)
    val encodedPassword = base32Encode(passwordData, LEGACY_BASE32)
    if (encodedNetworkId.length != 12 || encodedPassword.length != 13) {
        return ""
    }

    return listOf(
        encodedNetworkId.substring(0, 4),
        encodedNetworkId.substring(4, 8),
        encodedNetworkId.substring(8, 12),
        encodedPassword.substring(0, 5),
        encodedPassword.substring(5, 10),
        encodedPassword.substring(10, 13)
    ).joinToString("-").uppercase()
}

fun decodeConnectionCode(code: String): Pair<String, String> {
    val cleanCode = code.uppercase().replace(Regex("[^A-Z2-9]"), "")
    if (cleanCode.length != 25) return "" to ""

    val primaryResult = decodeConnectionCodeWithAlphabet(cleanCode, PRIMARY_BASE32)
    if (primaryResult.first.isNotEmpty() && primaryResult.second.isNotEmpty()) {
        return primaryResult
    }

    return decodeConnectionCodeWithAlphabet(cleanCode, LEGACY_BASE32)
}
